File: ArduinoPlatformio/02_demos/04 elementos/01 leds/01_Direct/03 RGB/04 setlevel/Platformio/lib/RoJoSwitch/RoJoSwitch.cpp
/*
  Nombre de la librería: RoJoSwitch.h
  Versión: 20210117
  Autor: Ramón Junquera
  Descripción:
    Gestión de un interruptor.
    Utiliza las resistencias PULL_UP internas.
    El interruptor debe ir conectado a tierra y al pin que lee su estado.
*/

#ifndef RoJoSwitch_cpp
#define RoJoSwitch_cpp

#include <RoJoSwitch.h>

//Constructor
//Se pasa el pin al que está conectado el pulsador. La otra pata del pulsador debe ir a GND
RoJoSwitch::RoJoSwitch(byte pin) { 
  _pinSwitch=pin; //Guardamos el pin al que está conectado el interruptor
  pinMode(_pinSwitch,INPUT_PULLUP); //Configuramos el pin de entrada con resistencia de PULLUP
}

//Comprueba el estado del interruptor
void RoJoSwitch::_check() {
  uint32_t now=millis(); //Tomamos nota de la hora actual
  if(digitalRead(_pinSwitch)!=_lastStatus) { //Si ha cambiado el estado del pulsador...
    _lastStatus=!_lastStatus; //Actualizamos es estado con el actual (lo cambiamos)
    if(now-_lastTime>delayTime) { //Si podemos aceptar una nueva pulsación...
      if(_lastStatus) _buttonReleased=true; //Si se acaba de soltar...indicamos que se ha soltado
      else _buttonPressed=true; //Si se acaba de pulsar...indicamos que se ha pulsado
    }
    _lastTime=now; //Tomamos nota del momento del cambio para esperar el tiempo indicado y evitar el efecto rebote
  }
}

//Indica si el interruptor ha sido pulsado
bool RoJoSwitch::pressed() {
  _check(); //Llamamos a la función que refresca el estado actual del interruptor
  if(_buttonPressed) { //Si ha sido pulsado...
    _buttonPressed=false; //...limpiamos el flag, porque ya ha sido consultado
    return true; //Devolvemos la respuesta
  }
  return false; //No ha sido pulsado
}

//Indica si el interruptor ha sido soltado
bool RoJoSwitch::released() {
  _check(); //Llamamos a la función que refresca el estado actual del interruptor
  if(_buttonReleased) { //Si ha sido soltado...
    _buttonReleased=false; //...limpiamos el flag, porque ya ha sido consultado
    return true; //Devolvemos la respuesta
  }
  return false; //No ha sido soltado
}

//Indica si el interruptor está siendo pulsado
bool RoJoSwitch::pressing() {
  _check(); //Llamamos a la función que refresca el estado actual del interruptor
  //Devolvemos el estado en el encuentra interruptor actualmente (el inverso porque es PULLUP)
  return !_lastStatus; 
}

#endif

File: Python/micropython/modulos/RoJoSwitch.py
#Autor: Ramón Junquera
#Fecha: 20221229

#Módulo RoJoSwitch
#Gestión de un interruptor
#Utiliza las resistencias PULL_UP internas
#El interruptor debe ir conectado a tierra y al pin que lee su estado

from machine import Pin,Signal
import time

class RoJoSwitch:
    def _check(self): #Comprueba el estado del pulsador
        now=time.ticks_ms() #Tomamos nota de la hora actual
        if self._pin.value()!=self._lastStatus: #Si ha cambiado el estado del pulsador...
            self._lastStatus=not self._lastStatus
            if now-self._lastTime>self.delayTime: #Si podemos aceptar una nueva pulsación...
                if self._lastStatus: #Si se acaba de soltar...
                    self._buttonReleased=True;
                else: #Si se acaba de pulsar...
                    self._buttonPressed=True
            _lastTime=now
    def __init__(self,pin):
        self._pin=Pin(pin,Pin.IN,Pin.PULL_UP)
        self._lastTime=0 #Momento del último cambio...ahora
        self._lastStatus=False #Último estado comprobado
        self._buttonPressed=False #Flag de si ha sido pulsado
        self._buttonReleased=False #Flag de si ha sido soltado
        self.delayTime=50 #Tiempo de espera en ms para evitar efecto rebote
        self._check()
        self._buttonPressed=self._buttonReleased=False #Inicialmente sin actividad
    def pressed(self): #Se ha pulsado?
        self._check()
        if self._buttonPressed: #Si se ha pulsado...
            self._buttonPressed=False #...Limpiamos el flag
            return True
        return False
    def released(self): #Se ha soltado?
        self._check()
        if self._buttonReleased: #Si se ha soltado...
            self._buttonReleased=False #Limpiamos el flag
            return True
        return False
    def pressing(self): #Se está pulsando?
        self._check()
        return not self._lastStatus
    def pin(self): #Devuelve el objeto Pin creado para el pulsador
        return self._pin

